zulip/zulip/__init__.py

```diff
@@ -1740,3 +1741,51 @@ def hash_util_decode(string: str) -> str:
     # Acknowledge custom string replacements in zulip/zulip's zerver/lib/url_encoding.py before unquoting.
     # NOTE: urllib.parse.unquote already does .replace('%2E', '.').
     return urllib.parse.unquote(string.replace(".", "%"))
+
+
+# Will be set to True during server CI tests for zulip of version 3.2 or prior to that
+USE_LEGACY_INTERFACE = os.environ.get("USE_LEGACY_INTERFACE", "False").lower() == "true"
+
+
+class LegacyInterfaceClient(Client):
+    """
+    This is a hack that we use to forcibly override the function signature of certain API bindings to make them support
+    older versions of zulip server in the CI.
+    """
+
+    @overload
+    def update_user_group_members(
+        self, user_group_id: int, group_data: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        ...
+
+    @overload
+    def update_user_group_members(self, group_data: Dict[str, Any]) -> Dict[str, Any]:
+        ...
+
+    def update_user_group_members(self, *args: Any, **kwargs: Any) -> Dict[str, Any]:
+        if isinstance(args[0], int) or isinstance(kwargs.get("user_group_id"), int):
+            return super().update_user_group_members(*args, **kwargs)
+
+        group_data = args[0] if len(args) > 0 else kwargs["group_data"]
+        modern_group_data = group_data.copy()
+        group_id = group_data["group_id"]
+        del modern_group_data["group_id"]
+        return super().update_user_group_members(group_id, modern_group_data)
+
+    def get_realm_filters(self) -> Dict[str, Any]:
+        """
+        Example usage:
+
+        >>> client.get_realm_filters()
+        {'result': 'success', 'msg': '', 'filters': [['#(?P<id>[0-9]+)', 'https://github.com/zulip/zulip/issues/%(id)s', 1]]}
+        """
+        # This interface was removed in 4d482e0ef30297f716885fd8246f4638a856ba3b
+        return self.call_endpoint(
+            url="realm/filters",
+            method="GET",
+        )
+
+
+if USE_LEGACY_INTERFACE:
+    Client = LegacyInterfaceClient  # type: ignore # We patch the original Client class to make it support older clients
```

Comment:
So, I'm thinking that we may need multiple legacy interface versions for different Zulip major releases if we want the tests to run against multiple versions; maybe we want to encode this as `LEGACY_INTERFACE_VERSION==3`, with the versions being the Zulip major release versions (E.g. 4 or 2.1 when the numbering scheme was different)?